sql/core/src/main/scala/org/apache/spark/sql/execution/SizeLimitingByteArrayDecoder.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.io.{ByteArrayInputStream, DataInputStream}
+
+import org.apache.spark.{SparkEnv, SparkException}
+import org.apache.spark.internal.{config, Logging}
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.util.Utils
+
+/**
+ * Provides methods for converting compressed byte arrays back to UnsafeRows.
+ * Additionally, can enforce a limit on the total, decoded size of all decoded UnsafeRows.
+ * Enforcing the limit is controlled via a sql config and if it is turned on the encoder will
+ * throw a SparkException when the limit is reached.
+ */
+private[spark] class SizeLimitingByteArrayDecoder(
+     nFields: Int,
+     sqlConf: SQLConf) extends Logging {
+  private var totalUncompressedResultSize = 0L
+  private val maxUncompressedResultSize = sqlConf.maxUncompressedResultSize
+
+  /**
+   * Decodes the byte arrays back to UnsafeRows and puts them into buffer.
+   */
+  def decodeUnsafeRows(bytes: Array[Byte]): Iterator[InternalRow] = {
+    val codec = CompressionCodec.createCodec(SparkEnv.get.conf)
+    val bis = new ByteArrayInputStream(bytes)
+    val ins = new DataInputStream(codec.compressedInputStream(bis))
+
+    new Iterator[InternalRow] {
+      private var sizeOfNextRow = ins.readInt()
+
+      override def hasNext: Boolean = sizeOfNextRow >= 0
+
+      override def next(): InternalRow = {
+        ensureCanFetchMoreResults(sizeOfNextRow)
+        val bs = new Array[Byte](sizeOfNextRow)
+        ins.readFully(bs)
+        val row = new UnsafeRow(nFields)
+        row.pointTo(bs, sizeOfNextRow)
+        sizeOfNextRow = ins.readInt()
+        row
+      }
+    }
+  }
+
+  private def ensureCanFetchMoreResults(sizeOfNextRow: Int): Unit = {
+    totalUncompressedResultSize += sizeOfNextRow
```

Comment:
We cannot check the actual data size before encoding for collect?

Reply:
Hm, we could calculate it before encoding and doing the collect of the data. But then we would need to do an extra collect that just collects the data sizes onto the driver and sums them and fails the job if the sum is too big? That would be bad performance-wise, right?

Reply:
But, after a driver finishes collecting all the compressed result data on memory (e.g., in executeCollect), is the current limiting approach is still useful for users? cc: @dongjoon-hyun @viirya

Reply:
Yea, I don't think it's useful. I think there's no point here since it's already collected.

Reply:
BTW, `spark.driver.maxResultSize` limits from executor side IIRC. So, that property is different.

Reply:
This is not more helpful than OOM can tell you. You still can't process the collected data...

Reply:
And, if you don't just collect one data, you can't set a good config to fit all cases.

Reply:
> This is not more helpful than OOM can tell you. You still can't process the collected data...

I meet this problem in SparkThriftServer. it's will make server hangs.

Reply:
This is useful when running many jobs on the same SparkContext. This protects one job from OOMing the JVM and crashing the SparkContext, killing all other jobs running on it.

Reply:
`spark.driver.maxResultSize` limits on the executor and on the driver @HyukjinKwon .
If a single result on an executor is > `spark.driver.maxResultSize` it fails the job. Also, `TaskSetManager` on the driver sums all the result sizes and fails the job if the result becomes bigger than that.
We can do the same thing for this new property if you think that would be better? Fail on the executor when encoding if it's already too big, otherwise fail on the driver.

Reply:
I've pushed a commit that also limits on the executors, similar to what `spark.driver.maxResultSize` does